subsys/mgmt/fmfu/src/mgmt_fmfu.c

```diff
@@ -0,0 +1,281 @@
+static int mgmt_firmware_upload(struct mgmt_ctxt *ctxt);
+static int unpack(struct mgmt_ctxt *ctxt, struct firmware_packet *packet, bool *whole_file_received);
+
+static const struct mgmt_handler mgmt_fmfu_handlers[] = {
+	[FMFU_MGMT_ID_GET_HASH] = {
+		.mh_read  = mgmt_get_memory_hash,
+		.mh_write = mgmt_get_memory_hash,
+	},
+	[FMFU_MGMT_ID_UPLOAD] = {
+		.mh_read  = NULL,
+		.mh_write = mgmt_firmware_upload
+	},
+};
+
+static struct mgmt_group fmfu_mgmt_group = {
+	.mg_handlers = (struct mgmt_handler *)mgmt_fmfu_handlers,
+	.mg_handlers_count = ARRAY_SIZE(mgmt_fmfu_handlers),
+	.mg_group_id = (MGMT_GROUP_ID_PERUSER + 1),
+};
+
+
+static int get_mgmt_err_from_modem_ret_err(int err)
+{
+	switch (err) {
+	case NRF_FMFU_RET_COMMAND_FAILED:
+		LOG_INF("RET_RPC_COMMAND_FAILED");
+		return MGMT_ERR_EMODEM_INVALID_COMMAND;
+	case NRF_FMFU_RET_COMMAND_FAULT:
+		LOG_INF("RET_RPC_COMMAND_FAULT");
+		return MGMT_ERR_EMODEM_FAULT;
+	default:
+		LOG_INF("BAD_STATE");
+		return MGMT_ERR_EBADSTATE;
+	}
+}
+
+int mgmt_fmfu_init(void)
+{
+	struct nrf_fmfu_digest digest;
+
+	LOG_INF("Initialized MFU mgmt module");
+	mgmt_register_group(&fmfu_mgmt_group);
+	int err = nrf_fmfu_init(&digest, NRF_FMFU_MODEM_BUFFER_SIZE,
+			(uint8_t *)0x20010000);
+	if (err != 0) {
+		LOG_ERR("nrf_fmfu_init failed: %d\n", err);
+		LOG_ERR("nrf_fmfu_init failed, errno: %d\n.", errno);
+		return err;
+	}
+	LOG_INF("Initialized MFU mgmt module done");
+	return err;
+}
+
+static int encode_response(struct mgmt_ctxt *ctx, uint32_t expected_offset)
+{
+	CborError err = 0;
+	int status = 0;
+
+	err |= cbor_encode_text_stringz(&ctx->encoder, "rc");
+	err |= cbor_encode_int(&ctx->encoder, status);
+	err |= cbor_encode_text_stringz(&ctx->encoder, "off");
+	err |= cbor_encode_uint(&ctx->encoder, expected_offset);
+
+	if (err != 0) {
+		return MGMT_ERR_ENOMEM;
+	}
+
+	return 0;
+}
+
+static int mgmt_firmware_upload(struct mgmt_ctxt *ctx)
+{
+
+	int ret;
+	int err;
+	bool whole_file_received;
+	static struct firmware_packet packet;
+	uint32_t next_expected_offset;
+
+	int modem_state = nrf_fmfu_modem_state_get();
+	static bool first_chunk = true;
+
+	LOG_INF("Modem state: %d", modem_state);
+	ret = unpack(ctx, &packet, &whole_file_received);
+	if (ret < 0) {
+		return ret;
+	}
+
+	next_expected_offset = ret;
+	LOG_INF("New offset: 0x%x", next_expected_offset);
+
+	if (packet.data_len > 0) {
+		if (first_chunk) {
+			err = nrf_fmfu_transfer_start();
+			if (err != 0) {
+				return get_mgmt_err_from_modem_ret_err(err);
+			}
+			first_chunk = false;
+		}
+
+		LOG_INF("target_addr: 0x%x", packet.target_address);
+		LOG_INF("Writing firmware packet with len:%d", packet.data_len);
+		err = nrf_fmfu_memory_chunk_write(packet.target_address,
+						  packet.data_len,
+						  &packet.data);
+		if (err != 0) {
+			LOG_INF("Error in writing data");
+			return get_mgmt_err_from_modem_ret_err(err);
+		}
+	}
+
+	if (whole_file_received) {
+		LOG_INF("whole file recived");
```

Comment:
I think this is not necessary, the offset being printed should be enough information about what has been passed to the modem.